### PR TITLE
B4B-10

### PR DESCRIPTION
**Pull request by OptimusOpus**
https://dltx.atlassian.net/jira/software/projects/ZIP/boards/60?selectedIssue=ZIP-10

Audit findings https://docs.google.com/document/d/1blREwoRMNTSMH7ivGImZm1E3LcfcxwRvZdA0bMqDkJA/edit#

### CODE REVIEW

File: contracts/B4REAL.sol [hunk 1]
```diff
@@ -125,18 +142,68 @@ contract B4REAL is ERC20, AccessControl {
...
+        return true;
+    }
+
+    function transferTax(
+        address sender,
+        uint256 amount
+    ) private {
+        address recipient = taxAddress;
+        require(sender != address(0), "ERC20: transfer from the zero address");
+        require(recipient != address(0), "ERC20: transfer to the zero address");
```
Comment (rellfy) [inline contracts/B4REAL.sol]: Should this really revert if the tax recipient is `address(0)`? Users wouldn't be able to transfer the token. Shouldn't the function `return` instead and not transfer out taxes?

File: contracts/B4REAL.sol [hunk 2]
```diff
@@ -125,18 +142,68 @@ contract B4REAL is ERC20, AccessControl {
...
+
+        super.transferFrom(sender, recipient, remainder);
+        return true;
+    }
+
+    function transferTax(
+        address sender,
+        uint256 amount
+    ) private {
+        address recipient = taxAddress;
```
Comment (rellfy) [inline contracts/B4REAL.sol]: I would check here if the amount is zero, and if so just return.
```suggestion
        if (amount == 0) return;
        address recipient = taxAddress;
```

File: contracts/B4REAL.sol [hunk 3]
```diff
@@ -1,24 +1,34 @@
 // SPDX-License-Identifier: MIT
-pragma solidity ^0.8.5;
+pragma solidity 0.8.15;
 
 import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
 import "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
 import "@openzeppelin/contracts/access/AccessControl.sol";
 
+import "hardhat/console.sol";
```
Comment (rellfy) [inline contracts/B4REAL.sol]: Do you want to remove this?

Comment (OptimusOpus) [reply contracts/B4REAL.sol]: Good spot! Cheers

File: contracts/B4REAL.sol [hunk 4]
```diff
@@ -9,16 +9,24 @@ contract B4REAL is ERC20, AccessControl {
...
     bytes32 public constant ADMIN_ROLE = keccak256("ADMIN_ROLE");
     bytes32 public constant OWNER_ROLE = keccak256("OWNER_ROLE");
 
     event ToggleWaiveFees(bool _status);
 
+    event SetTaxFee(uint256 _fee, uint256 _decimals);
+    event ExemptFromFee(address _account);
+    event IncludeInFee(address _account);
+    event UpdateB4REALTaxAddress(address _address);
+    event SetAdmin(address _account);
```
Comment (rellfy) [inline contracts/B4REAL.sol]: It seems like this is never emitted?

File: contracts/B4REAL.sol [hunk 5]
```diff
@@ -9,15 +9,21 @@ contract B4REAL is ERC20, AccessControl {
     using SafeERC20 for IERC20;
 
     address private taxAddress = 0xe3F078F80A530cCD3BbF221612dDca3B0724579D;
+    address public penddingOwner;
```
Comment (bazmatic) [inline contracts/B4REAL.sol]:
```suggestion
    address public pendingOwner;
```

File: contracts/B4REAL.sol [hunk 6]
```diff
@@ -46,6 +52,7 @@ contract B4REAL is ERC20, AccessControl {
     }
 
     constructor() ERC20("B4REAL", "B4RE") {
+        penddingOwner = msg.sender;
```
Comment (bazmatic) [inline contracts/B4REAL.sol]:
```suggestion
        pendingOwner = msg.sender;
```

File: contracts/B4REAL.sol [hunk 7]
```diff
@@ -12,12 +12,17 @@ contract B4REAL is ERC20, AccessControl {
     uint256 public taxFee;
     uint256 public taxFeeDecimals;
 
-    bool public waiveFees;
+    bool public waiveFees = false;
```
Comment (bitcoinbrisbane) [inline contracts/B4REAL.sol]: Any reason for this?  Its more expensive.

File: contracts/B4REAL.sol [hunk 8]
```diff
@@ -75,26 +85,30 @@ contract B4REAL is ERC20, AccessControl {
...
+        emit IncludeInFee(wallet);
     }
 
     /// @notice Updates the tax contract address
-    function updateB4REALTaxAddress(address newAddress) public onlyAdmin {
+    function updateB4REALTaxAddress(address newAddress) external onlyAdmin {
         require(taxAddress != newAddress, "New address cannot be the same");
+        require(newAddress != address(0), "The address cannot be the zero address");
         taxAddress = newAddress;
+        emit UpdateB4REALTaxAddress(newAddress);
```
Comment (bitcoinbrisbane) [inline contracts/B4REAL.sol]: I think this should be this way around, to prove it was set.
```suggestion
        emit UpdateB4REALTaxAddress(taxAddress);
```

File: contracts/B4REAL.sol [hunk 9]
```diff
@@ -117,26 +131,46 @@ contract B4REAL is ERC20, AccessControl {
...
+        public
+        override
+        returns (bool)
+    {
+        require(amount > 0, "The amount must be greater than 0");
+
+        uint256 remainder = amount;
+
+        // calculate the number of tokens the Tax should take
+        remainder = transferTax(sender, recipient, amount);
```
Comment (bitcoinbrisbane) [inline contracts/B4REAL.sol]: Not sure what we're doing here?  Why the asignment on LN150 https://github.com/dltxio/b4real-contracts/pull/2/files#diff-e436e94acae1c0f6ce625442eb79d2f19efc384cb235e97988b53d6553a2ca73R150

File: hardhat.config.ts [hunk 10]
```diff
@@ -70,11 +60,11 @@ export default {
   solidity: {
     compilers: [
       {
-        version: "0.8.7",
+        version: "0.8.15",
         settings: {
           optimizer: {
             enabled: true,
-            runs: 200
+            runs: 1
```
Comment (bitcoinbrisbane) [inline hardhat.config.ts]: Why the change?  Should be around the 200 mark.

File: tests/B4RealTest.spec.ts [hunk 11]
```diff
@@ -136,11 +165,172 @@ describe("B4REAL token tests", async () => {
...
+    const currentUser2Balance = await B4REAL.balanceOf(secondUserAddress);
+
+    const closingTaxBalance = Number(
+      ethers.utils.formatEther(await B4REAL.balanceOf(taxAddress))
+    );
+
+    // no tax should be applied so full amount should be sent to secondUserAddress
+    // and the tax address should have no balance change
+    expect(currentUser2Balance.toString()).to.equal(amount.toString());
+    expect(openingTaxBalance.toString()).to.equal(closingTaxBalance.toString());
```
Comment (bitcoinbrisbane) [inline tests/B4RealTest.spec.ts]: Cant we use a big number assert here?

File: tests/B4RealTest.spec.ts [hunk 12]
```diff
@@ -136,11 +165,172 @@ describe("B4REAL token tests", async () => {
...
+  });
+
+  it("Should test the transferFrom function and apply tax correctly", async () => {
+    await B4REAL.includeInFee(secondUserAddress);
+    const isWhitelisted = await B4REAL.whitelisted(secondUserAddress);
+    expect(isWhitelisted).to.equal(true);
```
Comment (bitcoinbrisbane) [inline tests/B4RealTest.spec.ts]: what about `.to.be.true`

File: tests/B4RealTest.spec.ts [hunk 13]
```diff
@@ -136,11 +165,172 @@ describe("B4REAL token tests", async () => {
...
+
+    const openingBalance = await B4REAL.balanceOf(secondUserAddress);
+
+    // Just the tokens from the previous test in this account
+    expect(openingBalance.toString()).to.equal(ethers.utils.parseEther("1"));
```
Comment (bitcoinbrisbane) [inline tests/B4RealTest.spec.ts]: Big Number assert?

File: tests/B4RealTest.spec.ts [hunk 14]
```diff
@@ -136,11 +165,172 @@ describe("B4REAL token tests", async () => {
...
+      method: "hardhat_impersonateAccount",
+      params: [secondUserAddress]
+    });
+
+    await B4REAL.connect(secondUserSigner).transferFrom(
+      userAddress,
+      secondUserAddress,
+      amount
+    );
+    await hre.network.provider.request({
```
Comment (bitcoinbrisbane) [inline tests/B4RealTest.spec.ts]:
```suggestion
    );
    
    await hre.network.provider.request({
```

File: tests/B4RealTest.spec.ts [hunk 15]
```diff
@@ -136,11 +165,172 @@ describe("B4REAL token tests", async () => {
...
+
+    const currentUser2Balance = await B4REAL.balanceOf(secondUserAddress);
+
+    const closingTaxBalance = Number(
+      ethers.utils.formatEther(await B4REAL.balanceOf(taxAddress))
+    );
+
+    // tax should be applied
+    expect(currentUser2Balance.toString()).to.equal("1900000000000000000");
+    expect(closingTaxBalance.toString()).to.equal("10.1");
```
Comment (bitcoinbrisbane) [inline tests/B4RealTest.spec.ts]: Big number assert here?

File: tests/B4RealTest.spec.ts [hunk 16]
```diff
@@ -136,11 +165,172 @@ describe("B4REAL token tests", async () => {
...
+    });
+    try {
+      await B4REAL.updateB4REALTaxAddress(
+        "0xe3F078F80A530cCD3BbF221612dDca3B0724579D"
+      );
+    } catch (error) {
+      expect(getRevertMessage(error)).to.equal(
+        "New address cannot be the same"
+      );
+    }
```
Comment (bitcoinbrisbane) [inline tests/B4RealTest.spec.ts:319]: Dont do this.  Use .to.be.revertedWith https://ethereum-waffle.readthedocs.io/en/latest/matchers.html#revert-with-message

**Review by bitcoinbrisbane — CHANGES_REQUESTED** [review]
Few things to tidy up.  Also I dont see an test on events.

File: tests/B4RealTest.spec.ts [hunk 17]
```diff
@@ -7,17 +7,21 @@ import { getRevertMessage } from "./utils";
```
Comment (bazmatic) [inline tests/B4RealTest.spec.ts:7]: If you use the `ethereum-waffle` middleware you can handle BigNumbers in your assertions.

```suggestion
import chai, { expect } from "chai";
import { solidity } from "ethereum-waffle";

chai.use(solidity);
```

Then you can go 
```
expect(myBigNumber).to.equal(anotherBigNumber);
```

File: tests/B4RealTest.spec.ts (hunk 13 above)
Comment (bazmatic) [reply tests/B4RealTest.spec.ts]: See my comment above about `ethereum-waffle` middleware for how to do that.

Comment (rellfy) [reply tests/B4RealTest.spec.ts]: Could do
```
expect(openingBalance.eq(ethers.utils.parseEther("1"))).to.be.true;
```

File: tests/B4RealTest.spec.ts (hunk 16 above)
Comment (OptimusOpus) [reply tests/B4RealTest.spec.ts:319]: I worked with Barry on this. For reasons we are still not sure about using `revertedWith` was breaking other tests

**Review by rellfy — APPROVED** [review]

**Review by bazmatic — APPROVED** [review]
